Include/Rover/Factor.hpp

```diff
@@ -0,0 +1,186 @@
+#ifndef ROVER_FACTOR_HPP
+#define ROVER_FACTOR_HPP
+#include <map>
+#include <optional>
+#include <type_traits>
+#include <unordered_map>
+#include <vector>
+
+namespace Rover {
+
+  //! Maps categories of a categorical variable to different dimensions.
+  /*!
+    \tparam T The type of the categorical variable with no operator < defined
+              or std::hash specialized.
+  */
+  template<typename T, typename = void>
+  class Factor {
+    public:
+
+      //! The type of the categorical variable.
+      using Type = T;
+
+      //! Registers a new category, if not already added.
+      /*!
+        \param category The category to add.
+        \details Amortized O(1) complexity.
+      */
+      void add_category(const Type& category);
+
+      //! Returns a unique dimension from 0 to (number of categories - 1) for
+      //! a given category. Returns nullopt if the category is not registered.
+      /*!
+        \param category The category.
+        \details O(N) complexity.
+      */
+      std::optional<std::size_t> get_dimension(const Type& category) const;
```

Comment:
Rename method to ```find_dimension```. It's a coding style rule (that I need to document) that ```get_XXX``` methods must basically be constant time and very cheap operations that can not fail.

Reply:
I renamed it, but I am not sure whether "find" really belongs here. From a class user's standpoint, "find" feels like accessing something that was explicitly stored at some point, like in a container. In this case, however, the dimension is calculated inside the class and is never displayed to the user until they request for it.
Yet, I cannot think of a better word for it, as something like "calculate"/"compute" implies heavy work and that the result should be cached, "access" means the result can be changed, etc.